keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -347,6 +349,30 @@ correct order.
 [enables]:https://github.com/kubernetes/kubernetes/blob/220498b83af8b5cbf8c1c1a012b64c956d3ebf9b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L703
 [filter]:#updating-storageversion
 
+#### Agreements
+
+1. Storageversion updates will be triggered by CRD create and update watch events
+2. We will perform async updates of storageversions such that we do not block writes of other unrelated APIs(for ex, crdA writes waiting for update of crdB's storageversion). These requests should be handled asynchronously
+3. We will block a storageversion update until the teardown of prev storage is finished
+   1. this means we will wait to update storageversion of a CRD until
+      1. old storage of the same CRD is deleted
+      2. pending CR writes using old storageversion of the same CRD are completed
+   2. We will do this to prevent publishing a newer storageversion while a pending CR write is still in progress. Otherwise, if the pending CR write finishes writing the CR using an old version, the storageversion API would not reflect that and the object will remain in etcd, encoded in an old version forever 
+4. We will block new CR writes for a CRD until we have published its latest storageversion. This is discussed more in the [limitations] section
+
+[limitations]:#Limitations 
+
+#### Limitations
+
+When a storageversion of a CRD is updated, we will ensure that all new CR writes:
+1. wait for the latest storageversion to be published
+1. wait for the CRD handler to be replaced
+
+This implies that we will take a write outage for the duration of the storageversion update. We will allow this for the following reasons:
+1. blocking CR writes till the SV update is finished is in-line with how we handle requests for built-in resources
+2. if we update the CRD handler to the new handler **before** a storageversion update completes, we risk a server crash - and objects maybe written in new version but storageversion API would not reflect that
```

Comment:
I don't think it's possible to always avoid this situation.  I think it's possible to always update the CRD handler before SV updates, I don't think we can guarantee atomicity for the CRD handler update and SV update (consider network partitions or process crashes happening at inconvenient times).  So while it seems nice to try to get the storageversion upgraded in the happy path, I think we need to also to tolerate the skew.

Reply:
IF we wan't to hold all writes indefinitely until SV is updated, we should spell out the requirements carefully.  What happens on apiserver startup when a CRD has been updated to a new storage version but the SV hasn't been updated yet.  What happens when updating the SV and the request times out?

In both cases, my understanding is that we loose availability of the CRD until the SV update completes.  we should call that out in the limitations.

Reply:
> What happens on apiserver startup when a CRD has been updated to a new storage version but the SV hasn't been updated yet. What happens when updating the SV and the request times out?

As discussed on chat, we will ensure that SVs are always reconciled when the server is started/restarted before we serve any CR writes.
I have updated this part to explicitly state the 2 scenarios where we will take an outage on CR write requests. 

About the SV update timing out, I have included a line saying we will reattempt the update a certain number of times before failing the SV udpate and consequently the dependednt CR writes. Does this look ok?